### DISCUSSION TITLE
下载任务时候，OnBeforePopup方法的疑问

### DISCUSSION THREAD

**Discussion by wangtao0505** [CefView/QCefView#401] [Q&A]
在页面上点击跳转链接或者点击下载链接的时候，我发现都会进OnBeforePopup方法。
此方法可以控制是否要打开新的窗口。那我如何判断是要跳转的链接是不是创建下载任务呢？

**Comment by L-Super** [comment]
可参考文档，当为下载文件时，会触发[onNewDownloadItem](https://cefview.github.io/QCefView/zh/docs/reference/QCefView#class_q_cef_view_1a666e93d6e9f9be7444f9b898f77c8292)，也可通过[CefWindowOpenDisposition](https://cefview.github.io/QCefView/zh/docs/reference/QCefView#class_q_cef_view_1a9963d810f8aa71b45b1b10f0abbe8787)判断对应类型

**Reply by wangtao0505** [reply]
onNewDownloadItem这个触发我理解。
如果下载的地址是以文件类型结尾，例如:http://xxxx.com/xxx.doc；http://xxxxx.com/xxxx.exe；……类似这样的链接，只会进onNewDownloadItem而不会进onBeforePopup方法(这样也不会创建新的窗口)。
但是如果下载链接是类似http://xxxxx.com?docId=xxxx这样的链接，就会进onBeforePopup。而且我加断点看[CefWindowOpenDisposition]的值一直等于3

**Comment by L-Super** [comment]
第二种方式，会无法下载文件吗？

**Reply by wangtao0505** [reply]
可以下载，但是需要打开要跳转的URL地址。新窗口或者本窗口打开都行。
而且我试了，如果在选择好本地保存路径之后，下载开始前关闭了这个url窗口，下载会失败。

**Reply by wangtao0505** [reply]
试了一下CEF官方提示的可执行文件，打开这样的下载链接，也会弹出一个新的窗口。
我现在的方案是：下载的时候，新建一个窗口，并且记录下当前这个窗口的url地址。在onUpdateDownloadItem方法中当下载结束或者取消下载的时候，根据item->originalUrl();的值，将刚刚打开的新窗口关闭。
虽然下载的时候看着有些别扭，但还算可以接收。

以上方案产生了一个新的问题，CCefClientDelegate::onBeforeDownload的这个地方：QMetaObject::invokeMethod( pCefViewPrivate_, [=]() { pCefViewPrivate_->onNewDownloadItem(item, suggestedFileName); }, c)；偶发阻塞了。应该是pCefViewPrivate_里面某个地方跟这块死锁了。将ConnectionType改为QueuedConnection之后，还没复现。

**Reply by tishion** [reply]
弹出新窗口是因为你的前端页面实现的逻辑就是要弹出。跟CEF无关。

**Reply by wangtao0505** [reply]
是的，但是即便不做新窗口，如果我说的这种链接，还行需要打开对应的URL的。如果不做新窗口，就只能覆盖当前窗口了。下载结束之后再goback
